File: scripts/send_to_cloud.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# 距離を読む関数  
def read_distance():
    # 必要なライブラリのインポート・設定 
    import time
    import RPi.GPIO as GPIO
    GPIO.setwarnings(False)
    
    # 使用するピンの設定
    GPIO.setmode(GPIO.BOARD)
    TRIG = 11 # ボード上の11番ピン(GPIO17)
    ECHO = 13 # ボード上の13番ピン(GPIO27)
    
    # ピンのモードをそれぞれ出力用と入力用に設定 
    GPIO.setup(TRIG,GPIO.OUT)
    GPIO.setup(ECHO,GPIO.IN)
    GPIO.output(TRIG, GPIO.LOW)
    time.sleep(0.3)
    
    # TRIGに短いパルスを送る
    GPIO.output(TRIG, GPIO.HIGH)
    time.sleep(0.00001)
    GPIO.output(TRIG, GPIO.LOW)

    # ECHO ピンがHIGHになるのを待つ
    while GPIO.input(ECHO) == GPIO.LOW:
        signaloff = time.time()
     
    # ECHO ピンがLOWになるのを待つ
    while GPIO.input(ECHO) == GPIO.HIGH:
        signalon = time.time()

    # 時刻の差から、物体までの往復の時間を求め、距離を計算する 
    timepassed = signalon - signaloff
    distance = timepassed * 17000
    return distance
    GPIO.cleanup()

# IMSIの取得
import requests
import json

print "- メタデータサービスにアクセスして IMSI を確認中 ...",
subscriber=json.loads(requests.get('http://metadata.soracom.io/v1/subscriber').text)
imsi = subscriber['imsi']
print imsi

# 閾値の設定: 距離 10cm 以内が３回続いた場合にイベント
threshold_distance = 10
threshold_count = 3
status = 'out' # 何もない時は out 、何かある時は in
duration = 0
count = 0

print "- 条件設定"
print "障害物を %d cm 以内に %d 回検知したらクラウドにデータを送信します" % (threshold_distance, threshold_count )
print "センサーを手で遮ったり、何か物を置いてみたりしてみましょう"

import time
status_changed_at = time.time()

import datetime
from elasticsearch import Elasticsearch
es = Elasticsearch('beam.soracom.io:18080')

print "- 準備完了"

while True:
    distance = read_distance() # 距離の取得

    if status == 'out':
        if distance <= threshold_distance:
            count+=1
            print "距離(cm): %.1f <= %1.f , 回数: %d / %d" % (distance, threshold_distance, count, threshold_count)
        else:
            count=0
    else:
        if distance > threshold_distance:
            count+=1
            print "距離(cm): %.1f > %1.f , 回数: %d / %d" % (distance, threshold_distance, count, threshold_count)
        else:
            count=0

    if count >= threshold_count:
        status = 'out' if status == 'in' else 'in'
        duration = time.time() - status_changed_at
        status_changed_at = time.time()
        print "- ステータスが '%s'(%s) に変化しました" % (status, '何か物体がある' if status == 'in' else '何も物体がない')
        print "- Beam 経由でデータを送信します"
        print es.index(index="sensor", doc_type="event", body={"imsi":imsi, "status":status, "timestamp":datetime.datetime.utcnow().strftime('%Y-%m-%dT%H:%M:%S.%fZ'), "duration":duration})
        count=0

File: scripts/send_to_cloud.rb
#!/usr/bin/env ruby
require './GPIO'
require 'json'
require 'open-uri'
require 'elasticsearch'

GPIO.reset(17,27)
GPIO.export(17, 'out')
GPIO.export(27, 'in')

def read_distance
  GPIO.write(17, 1)
  sleep 0.00001
  GPIO.write(17, 0)
  
  edge = GPIO::Edge.new(27)
  edge.read
  edge.wait(1)
  s = Time.now
  edge.wait(1)
  e = Time.now
  edge.close
  return (e.to_f - s.to_f) * 17000
end

begin
  print "- メタデータサービスにアクセスして IMSI を確認中 ... "
  subscriber=JSON.parse(open('http://metadata.soracom.io/v1/subscriber').read)
  imsi=subscriber['imsi']
  puts imsi
rescue => e
  abort "ERROR: メタデータサービスが有効になっていない可能性があります"
end

# 閾値の設定: 距離 10cm 以内が３回続いた場合にイベント
threshold_distance = 10
threshold_count = 3
status = 'out' # 何もない時は out 、何かある時は in
status_change_at = Time.now
duration = 0
count = 0

puts <<EOF
- 条件設定
障害物を #{threshold_distance} cm 以内に #{threshold_count} 回検知したらクラウドにデータを送信します
センサーを手で遮ったり、何か物を置いてみたりしてみましょう
EOF

# ElasticSearch クライアント設定
es = Elasticsearch::Client.new hosts:[ {host: 'beam.soracom.io', port: '18080'} ]

puts "- 準備完了"

while true
  start_time = Time.now
  distance = read_distance # 距離の取得

  if status == 'out'
    if distance <= threshold_distance # 障害物を検知 
      count+=1
      printf("距離(cm) %.1f <= %.1f , 回数: %d / %d\n", distance, threshold_distance, count, threshold_count)
    else
      count=0
    end
  else 
    if distance > threshold_distance # 障害物がなくなったのを検知
      count+=1
      printf("距離(cm) %.1f > %.1f , 回数: %d / %d\n", distance, threshold_distance, count, threshold_count)
    else
      count=0
    end
  end

  # 状態が変化した
  if count >= threshold_count
    status = (status == 'out')? 'in':'out'
    duration = Time.now - status_change_at
    puts "- ステータスが '#{status}' (#{status == 'in'? '何か物体がある':'何も物体がない'}) に変化しました"
    puts "- Beam 経由でデータを送信します"
    puts es.index index: 'sensor', type: 'event', body:{'imsi':imsi, 'status':status, 'duration':duration, 'timestamp':Time.now.to_i*1001}
    count = 0
  end

  # １秒に１回実行するためのウェイトを入れる
  if Time.now < start_time+1
    sleep start_time+1-Time.now
  end
end
